walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -11,13 +11,130 @@
+import java.util.Set;
+ 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // create a full day of free timeslots
+    List<TimeRange> fullDay = new ArrayList<TimeRange>();
+    fullDay.add(TimeRange.WHOLE_DAY);
+    
+    // cannot modify arraylist in an enhanced for loop, must create separate arraylists
+    List<TimeRange> freeTimes = new ArrayList<TimeRange>();
+    List<TimeRange> blockedTimes = new ArrayList<TimeRange>();
+
+
+    Collection<String> attendees = request.getAttendees();
+    Set<String> eventAttendees;
+    
+  for(Event event: events){
+    eventAttendees = event.getAttendees();
+    boolean isOverlappedAttendee = false;
+    for(String attendee:eventAttendees){
+      if(attendees.contains(attendee)){
+        isOverlappedAttendee = true;
+        // add a break to terminate loop early if at least one attendee is present at event
+        break; 
+        }
+      }
+      if(isOverlappedAttendee){
+        TimeRange blocked = event.getWhen();
+        for(TimeRange timeslot: fullDay){
+          if(blocked.overlaps(timeslot)){
+            // case one: event is fully enclosed within a free timeslot
+            if(blocked.start() > timeslot.start() && blocked.end() < timeslot.end()){
+              freeTimes.add(TimeRange.fromStartEnd(timeslot.start(), blocked.start(),false));
+              freeTimes.add(TimeRange.fromStartEnd(blocked.end(),timeslot.end(),false));
+            }
+            // case two: event overlaps end of free timeslot
+            else if(blocked.start() >= timeslot.start() && blocked.end() >= timeslot.end()){
+              if(blocked.start()  != timeslot.start()){
+                freeTimes.add(TimeRange.fromStartEnd(timeslot.start(), blocked.start(),false));
+              }
+            }
+            // case three: event overlaps start of free timeslot
+            else if(blocked.start() <= timeslot.start() && blocked.end() >= timeslot.start()){
+              if(blocked.end()  != timeslot.end()){
+                freeTimes.add(TimeRange.fromStartEnd(blocked.end(),timeslot.end(),false));
+              }
+            }
+            blockedTimes.add(timeslot);
+          }
+        }
+        fullDay.addAll(freeTimes);
+        fullDay.removeAll(blockedTimes);
+        // must clear everytime so that blocked events don't get added back
+        freeTimes.clear();
+        blockedTimes.clear();
+      }
+    }
+        
+    // filter through available times to find sufficient durations
+    for(TimeRange available: fullDay){
+      if(available.duration() < request.getDuration()){
+          blockedTimes.add(available);
+      }
+    }
+    fullDay.removeAll(blockedTimes);
+    
+    // try to insert optional attendees into existing timeslots 
+    attendees = request.getOptionalAttendees();
+    for(Event event: events){
```

Comment:
This for loop section seems very similar to the one above it. Do you think you could move this to a separate helper function and call it twice with both optional and mandatory meeting attendees

Reply:
I've now simplified the second loop so they are no longer very similar, but I've gone ahead and move the nested ifs into a helper function as well.